src/msg/producer/config/writer.go

```diff
@@ -239,3 +255,80 @@ func (c *WriterConfiguration) setRetryOptions(
 	}
 	return opts, nil
 }
+
+// WatchGracefulClose watches for graceful close setting updates from KV.
+func (c *WriterConfiguration) WatchGracefulClose(
+	cs client.Client,
+	gracefulClose *bool,
+	logger *zap.Logger,
+) {
+	const defaultGracefulClose = false
+
+	// If no key is configured, skip watching and use default
+	if c.MessageWriterGracefulCloseKey == "" {
+		logger.Info("message writer graceful close key not configured, using default value",
+			zap.Bool("defaultGracefulClose", defaultGracefulClose))
+		return
+	}
+
+	kvOpts, err := c.MessageWriterGracefulCloseKVOverride.NewOverrideOptions()
+	if err != nil {
+		logger.Error("unable to create kv config options for graceful close", zap.Error(err))
+		return
+	}
+
+	store, err := cs.Store(kvOpts)
+	if err != nil {
+		logger.Error("unable to create kv store for graceful close", zap.Error(err))
+		return
+	}
+
+	var (
+		gracefulCloseKey = c.MessageWriterGracefulCloseKey
+		gracefulCloseCh  <-chan struct{}
+	)
+
+	utilOpts := kvutil.NewOptions().SetLogger(logger)
+
+	// Retrieve initial value from KV
+	initialValue := defaultGracefulClose
+	value, err := store.Get(gracefulCloseKey)
+	if err == nil {
+		initialValue, err = kvutil.BoolFromValue(value, gracefulCloseKey, defaultGracefulClose, utilOpts)
+	}
+	if err != nil {
+		logger.Error("unable to retrieve graceful close setting from kv", zap.Error(err))
+	}
+	logger.Info("current message writer graceful close setting", zap.Bool("gracefulClose", initialValue))
+	*gracefulClose = initialValue
+
+	// Set up watch
+	gracefulCloseWatch, err := store.Watch(gracefulCloseKey)
+	if err != nil {
+		logger.Error("unable to watch message writer graceful close setting", zap.Error(err))
+		return
+	}
+	gracefulCloseCh = gracefulCloseWatch.C()
+
+	// Watch for updates
+	go func() {
```

Comment:
There is no graceful termination of this goroutine since no one is closing the channel right ? On termination of aggregator, this goroutine will be ungracefully terminated. Let's keep a reference of spawned goroutines to ensure clean service shutdown ?